server/app/controllers/applicant/ApplicantProgramBlocksController.java

```diff
@@ -129,6 +150,147 @@ public CompletionStage<Result> review(
+        applicantStage
+            .thenComposeAsync(
+                v -> checkApplicantAuthorization(profileUtils, request, applicantId),
+                httpExecutionContext.current())
+            .toCompletableFuture();
+
+    CompletableFuture<ReadOnlyApplicantProgramService> applicantProgramServiceCompletableFuture =
+        applicantStage
+            .thenComposeAsync(
+                v -> applicantService.getReadOnlyApplicantProgramService(applicantId, programId),
+                httpExecutionContext.current())
+            .toCompletableFuture();
+
+    return CompletableFuture.allOf(
+            applicantAuthCompletableFuture, applicantProgramServiceCompletableFuture)
+        .thenApplyAsync(
+            (v) -> {
+              ReadOnlyApplicantProgramService roApplicantProgramService =
+                  applicantProgramServiceCompletableFuture.join();
+              Optional<Block> block = roApplicantProgramService.getBlock(blockId);
+              Optional<String> maybeAddressJson = request.session().get(ADDRESS_JSON_SESSION_KEY);
+
+              if (block.isEmpty()) {
+                return notFound();
+              }
+
+              if (maybeAddressJson.isEmpty()) {
+                throw new RuntimeException("Address JSON missing");
+              }
+
+              try {
+                Block thisBlockUpdated = block.get();
+                ObjectMapper mapper = new ObjectMapper();
+                JsonNode rootNode = mapper.readTree(maybeAddressJson.get());
+
+                ImmutableList.Builder<AddressSuggestion> suggestionBuilder =
+                    ImmutableList.builder();
+                for (JsonNode node : rootNode) {
+                  suggestionBuilder.add(AddressSuggestion.builder().fromJson(node));
+                }
+
+                ImmutableList<ApplicantQuestion> questions = thisBlockUpdated.getQuestions();
+
+                Optional<ApplicantQuestion> addressQuestionMaybe =
+                    questions.stream()
+                        .filter(
+                            m ->
+                                m.getQuestionDefinition().isAddress()
+                                    && m.isAddressCorrectionEnabled())
+                        .findFirst();
+
+                if (addressQuestionMaybe.isPresent()) {
+                  AddressQuestion addressQuestion =
+                      addressQuestionMaybe.get().createAddressQuestion();
+                  DynamicForm form = formFactory.form().bindFromRequest(request);
+                  String selectedAddress =
+                      form.get(AddressCorrectionBlockView.SELECTED_ADDRESS_NAME);
+
+                  Optional<AddressSuggestion> suggestionMaybe =
+                      suggestionBuilder.build().stream()
+                          .filter(x -> x.getSingleLineAddress().equals(selectedAddress))
+                          .findFirst();
+
+                  Map<String, String> map = new HashMap<>(Map.of());
```

Comment:
done

Reply:
Suggestion (blocking): If the address correction fields were on the form itself, it would be simpler to get the form data like is done in the `update` method:
```
DynamicForm form = formFactory.form().bindFromRequest(request);
              ImmutableMap<String, String> formData = cleanForm(form.rawData());
```

Reply:
Please ignore my last comment, since the fields won't get added to the DOM.